Make image size limit configurable, expose to avifdec

joedrago opened the pull request:
This should resolve #263 .

@wantehchang:

* Should we make the default 0 (no limit) and manually set the limit in our fuzzer (and possibly Chrome), or should we keep the constant `AVIF_MAX_IMAGE_SIZE` (perhaps renamed to have DEFAULT somewhere in it?) and force people to opt into no limit?

* Does libaom have a similar size limit we can use? It'd be nice to not have to caveat "dav1d only" in various places.

wantehchang commented on include/avif/avif.h:
```diff
@@ -701,6 +705,11 @@ typedef struct avifDecoder
 
+    // This represents the maximum size of a image (in pixel count) that the underlying AV1 decoder
+    // should attempt to decode. It defaults to AVIF_MAX_IMAGE_SIZE, and can be set to 0 to disable
+    // the limit. Currently supported codecs: dav1d.
```
1. It would be good to note that other AV1 codecs may have a non-configurable frame size limit. That's the case for libgav1 and libaom.

For libgav1, the frame size limit is hardcoded in the source code, currently INT32_MAX:

```
  if (frame_header_.upscaled_width > INT32_MAX / frame_header_.height) {
    LIBGAV1_DLOG(ERROR, "Frame dimensions too big: width=%d height=%d.",
                 frame_header_.width, frame_header_.height);
    return false;
  }
```

For libaom, the frame size limit is a build opton, consisting of three cmake flags. Here is how libaom's source code uses the three corresponding C-preprocessor macros:

```
static AOM_INLINE void resize_context_buffers(AV1_COMMON *cm, int width,
                                              int height) {
#if CONFIG_SIZE_LIMIT
  if (width > DECODE_WIDTH_LIMIT || height > DECODE_HEIGHT_LIMIT)
    aom_internal_error(&cm->error, AOM_CODEC_CORRUPT_FRAME,
                       "Dimensions of %dx%d beyond allowed size of %dx%d.",
                       width, height, DECODE_WIDTH_LIMIT, DECODE_HEIGHT_LIMIT);
#endif
```

Note that the clients of libaom cannot change the size limit at run time.

2. Clarify whether imageSizeLimit applies to grid image size.

wantehchang commented on include/avif/avif.h:
```diff
@@ -76,6 +76,10 @@ typedef int avifBool;
 
+// A maximum image size to avoid out-of-memory errors or integer overflow in
+// (32-bit) int or unsigned int arithmetic operations.
+#define AVIF_MAX_IMAGE_SIZE (16384 * 16384)
```
Your suggestion of adding DEFAULT to the macro's name is good.

wantehchang reviewed (changes requested):
Joe: I suggest we be secure by default and force people to opt into no limit.

joedrago commented on include/avif/avif.h:
```diff
@@ -701,6 +705,11 @@ typedef struct avifDecoder
+    // This represents the maximum size of a image (in pixel count) that the underlying AV1 decoder
+    // should attempt to decode. It defaults to AVIF_MAX_IMAGE_SIZE, and can be set to 0 to disable
+    // the limit. Currently supported codecs: dav1d.
+    uint32_t imageSizeLimit;
```
`imageSizeLimit` is not a member of `avifImage`, so bullets 1 and 2 don't apply here, right?

wantehchang replied:
Sorry about my confusion. I checked the wrong function. Bullet 1 should read:

> 1. In src/read.c, avifDecoderCreate() needs to initialize imageSizeLimit to the default (AVIF_MAX_IMAGE_SIZE).

Bullet 2 is correct. I was referring to the following code in avifParseImageGridBox():

```
    if ((grid->outputWidth == 0) || (grid->outputHeight == 0) || (grid->outputWidth > (AVIF_MAX_IMAGE_SIZE / grid->outputHeight))) {
        return AVIF_FALSE;
    }
```

joedrago replied:
Ah, I see. That is a good question.

Do we want `imageSizeLimit` to be something solely enforced by the underlying AV1 decoders, or do we want to explicitly check these limits even at the container level? For example, if the `ispe` box returned something that exceeds `imageSizeLimit`, do we bail out immediately? Do we want a new error of `AVIF_RESULT_IMAGE_TOO_LARGE`?

wantehchang replied:
imageSizeLimit serves two purposes:

1. To help prevent integer arithmetic overflow in libavif client's code. For this purpose, we want to explicitly check the image size limits even at the container level. (I believe we are not doing that now.) I think we should add a new error of `AVIF_RESULT_IMAGE_TOO_LARGE`.
2. oss-fuzz tests fail if we allocate too much memory from the heap. (oss-fuzz doesn't care whether we handle out-of-memory errors properly.) For this purpose, we don't need to check the image size limits at the container level, but we do need to check the grid image sizes (because we will call `avifImageAllocatePlanes()` in `avifDecoderDataFillImageGrid()`), and we also need the underlying AV1 decoders to enforce the same size limits.

joedrago commented:
Alright, I've renamed `AVIF_MAX_IMAGE_SIZE` to `AVIF_DEFAULT_MAX_IMAGE_SIZE`, set it in `avifDecoder`, added `AVIF_RESULT_IMAGE_TOO_LARGE` and honor it in both grid and ispe cases, did an audit pass on `reformat.c` and `alpha.c` to ensure we were using `uint32_t` for our pixel indexing math local variables, and tweaked some comments. When you get a chance, please re-audit the PR fresh.

joedrago commented:
Hm, I realized right as I typed that that `uint32_t` is insufficient for indexing if ispe really allows for `uint32_t` width and heights. Yuck.

wantehchang commented on src/read.c:
```diff
@@ -1208,15 +1208,18 @@ static avifBool avifParseImageGridBox(avifImageGrid * grid, const uint8_t * raw,
+    if ((grid->outputWidth == 0) || (grid->outputHeight == 0) || (avifROStreamRemainingBytes(&s) != 0)) {
+        return AVIF_RESULT_INVALID_IMAGE_GRID;
+    }
+    if (imageSizeLimit && (grid->outputWidth > (imageSizeLimit / grid->outputHeight))) {
```
If imageSizeLimit is equal to 0, we still need to make sure grid->outputWidth * grid->outputHeight does not overflow size_t. We can either check that here or when we allocate the plane buffers.

wantehchang commented on src/reformat.c:
```diff
@@ -307,10 +307,10 @@ avifResult avifImageRGBToYUV(avifImage * image, const avifRGBImage * rgb)
+                const uint32_t chromaShiftX = 1;
+                const uint32_t chromaShiftY = 1;
+                uint32_t uvI = outerI >> chromaShiftX;
+                uint32_t uvJ = outerJ >> chromaShiftY;
```
Here we need to make sure the multiplication uvJ * yuvRowBytes[AVIF_CHAN_U] at line 315 below is done using the size_t type and also make sure that multiplication does not overflow size_t.

It is best to check for potential overflow on entry to this function.

By now it should be clear why we (and Chrome) set AVIF_MAX_IMAGE_SIZE to 2^28, so that we can perform these arithmetic operations using the 32-bit 'int' or 'uint32_t' type freely.
